### PR TITLE
Enhance tests output

### PR DESCRIPTION
**Pull request by RussKie**
Contributes to #7702

### CODE REVIEW

**Comment by RussKie** [comment]
Please ignore the first three commits, they are used for some build failures.

File: appveyor.yml [hunk 1]
```diff
@@ -102,7 +100,7 @@ test_script:
             }
         }
     }
-    if ($testErrorsTotal -ne 0  -and  $env:IGNORE_TEST_RESULT -ne "true") {
```
Comment (mstv) [inline appveyor.yml]: Do you want to remove this variable or just for testing -- WIP?

File: appveyor.yml [hunk 2]
```diff
@@ -80,15 +80,15 @@ test_script:
...
-                #$result[0].Value
+
                 $total  = $result[1].Value
                 $passed = $result[4].Value
                 $failed = $result[5].Value
                 $notPassed = $total - $passed
                 $testErrors = $failed
                 $testName = $resultFilePath.Split('\\')[-3]
                 $artifactName = "TestResult#${testRun}.${testErrors}err.${testName}"
-                Push-AppveyorArtifact $resultFilePath -FileName "${artifactName}.xml"
```
Comment (mstv) [inline appveyor.yml]: I agree, the artifacts have become quite noisy.
At least on error, this file should be kept because it contains the most complete information.
(The best solution IMO would be one `.xml` with all the test results.)
An alternative could be to zip them into one archive -- with the drawback that it cannot be opened directly.

**Comment by mstv** [comment]
> NUnit : error : Tests failed: C:\projects\gitextensions\IntegrationTests\UI.IntegrationTests\UI.IntegrationTests.csproj [net461|x64] [C:\projects\gitextensions\IntegrationTests\UI.IntegrationTests\UI.IntegrationTests.csproj]
Build FAILED.

The CI build should not fail before all tests have been run.

File: appveyor.yml [hunk 3]
```diff
@@ -80,25 +79,31 @@ test_script:
...
+
+                Write-Host "Pass#${testRun}: passed: $passed, failed: $failed, not passed: $notPassed, total: $total"
```
Comment (mstv) [inline appveyor.yml]:
```suggestion
                Write-Host "Pass #${testRun}: passed: $passed, failed: $failed, not passed: $notPassed, total: $total"
```

File: appveyor.yml [hunk 4]
```diff
@@ -80,25 +79,31 @@ test_script:
...
+
                 if ($notPassed -ne 0) {
                     $logFile = Get-ChildItem -recurse artifacts\$testname*.log
                     Get-Content $logFile
                     if ($testErrors -ne 0) { Push-AppveyorArtifact $logFile -FileName "${artifactName}.log" }
                 }
                 $testErrorsTotal += $testErrors
+                Write-Host "Pass#${testRun}: testErrorsTotal: $testErrorsTotal"
```
Comment (mstv) [inline appveyor.yml]:
```suggestion
                Write-Host "Pass #${testRun}: testErrorsTotal: $testErrorsTotal"
```

File: appveyor.yml [hunk 5]
```diff
@@ -80,25 +79,31 @@ test_script:
...
                 $testErrors = $failed
                 $testName = $resultFilePath.Split('\\')[-3]
                 $artifactName = "TestResult#${testRun}.${testErrors}err.${testName}"
-                Push-AppveyorArtifact $resultFilePath -FileName "${artifactName}.xml"
+
+                Write-Host "Pass#${testRun}: passed: $passed, failed: $failed, not passed: $notPassed, total: $total"
+
                 if ($notPassed -ne 0) {
                     $logFile = Get-ChildItem -recurse artifacts\$testname*.log
                     Get-Content $logFile
```
Comment (mstv) [inline appveyor.yml]: This file is printed by the Arcade script now.
```suggestion
```

File: appveyor.yml (hunk 1 above)
Comment (RussKie) [reply appveyor.yml]: I've been mulling over it over night.

I want to remove all of this plumbing for the following reasons:
- Both master and release/*  branches are release branches, and shouldn't contain any "for testing purposes" infra. The latter belongs to pull-requests/feature branches.
- We must fail a build if either unit or integration tests fail. If unit tests fail - we have regressed, it must be fixed, no point running integration tests.

If you need this plumbing for testing purposes, please push these changes to branches on the main trunk instead of your fork. This way changes to appveyor.yml will be executed.

File: appveyor.yml (hunk 2 above)
Comment (RussKie) [reply appveyor.yml]: I think I may have asked before, but what's the benefit of these xml for user? 
The main reason we have them for opencover/appveyor to report the stats. As a user a I look at the *.log files.
Merging xmls together is a time consuming operation which will result in a tens of thousands of lines of xml....

File: appveyor.yml (hunk 3 above)
Comment (RussKie) [reply appveyor.yml]: These prints are for debugging purposes only, I'll remove them in the final version.
It took me awhile to realise test failures were ignore at the AppVeyor level... 🤦‍♂

File: appveyor.yml (hunk 5 above)
Comment (RussKie) [reply appveyor.yml]: Ah! Thank you, will clean up

**Comment by RussKie** [comment]
@mstv I have cleaned the code. I think it is ready.
I will remove commits other than 92641e8c3b8b96a9ebbd0a2929fdd47b01d1c6fd before the merge.

**Comment by pmiossec** [comment]
@RussKie 

>I will remove commits other than 92641e8 before the merge.

I think that you could remove it because #7745 succeed in making the build fail

File: appveyor.yml [hunk 6]
```diff
@@ -1,23 +1,15 @@
...
-  # The following entries define the default values of the environment variables.
-  # They can be overridden in AppVeyor | Settings | Environment.
-  matrix:
-  - IdeVersion: VS2019
-  SKIP_PAUSE: TRUE
-  ARCHIVE_WITH_PDB: TRUE
-  GE_TEST_SLEEP_SECONDS_ON_HANG: 0
-  GE_TEST_LAUNCH_DEBUGGER_ON_HANG: 0
-  TEST_RUN_COUNT: 1
-  IGNORE_TEST_RESULT: FALSE
```
Comment (mstv) [inline appveyor.yml]: Why did you remove all AppVeyor variables?
These four still make sense to me.

File: appveyor.yml [hunk 7]
```diff
@@ -1,23 +1,15 @@
...
-os: Visual Studio 2019
-
 branches:
   except:
     - configdata
     - gh-pages
 
-environment:
-  # The following entries define the default values of the environment variables.
-  # They can be overridden in AppVeyor | Settings | Environment.
```
Comment (mstv) [inline appveyor.yml]: This comment is important IMO if we keep this section.

File: appveyor.yml (hunk 6 above)
Comment (RussKie) [reply appveyor.yml]: They are useful for debug builds, but we shouldn't have them in the main build.
The master and the release builds aren't meant to be used as test grounds. For tests we can create branches off the main trunk and modify appveyor.yml as required to complete tests.

File: appveyor.yml (hunk 7 above)
Comment (RussKie) [reply appveyor.yml]: This whole section is gone...

File: appveyor.yml (hunk 6 above)
Comment (mstv) [reply appveyor.yml]: Git does not support this in a user-friendly way. This will end up in lots of merges, rebases and merge conflicts.
I don't need it as AppVeyor/Environment variables -- I was told to use them.
But the "extended test infrastructure" should be available and maintained in the main trunk.

BTW: The build script seems to *not* set the exit code.

File: appveyor.yml (hunk 7 above)
Comment (RussKie) [reply appveyor.yml]: We can't have it without something defined
image

File: appveyor.yml (hunk 6 above)
Comment (RussKie) [reply appveyor.yml]: > I don't need it as AppVeyor/Environment variables -- I was told to use them.

I suggested to use AppVeyor env variables as a way to dynamically change behaviours, if that's what you're referring to. It is certainly a better way than hardcoding magic numbers in the common infrastructure.

> But the "extended test infrastructure" should be available and maintained in the main trunk.

Why?
The master and release branches must be stable. If we have issues with tests (like now), we should create a feature branch in the main trunk and tweak its pipeline to suit (e.g. run tests multiple times, ignore errors, etc). Once the work is complete, the fix is submitted to the master, and all temporary tweaks are discarded.

Comment (RussKie) [reply appveyor.yml]: > BTW: The build script seems to _not_ set the exit code.

Could you please elaborate?

In a typical flow, `Build` function is invoked and returns the code:
https://github.com/gitextensions/gitextensions/blob/c110ab60f1ab3075b3e8e8213a1990da698fd006/Build/Build.ps1#L75-L91

If there is an exception, it returns here:
https://github.com/gitextensions/gitextensions/blob/c110ab60f1ab3075b3e8e8213a1990da698fd006/Build/Build.ps1#L113-L116

What am I missing?

Comment (mstv) [reply appveyor.yml]: > Once the work is complete, the fix is submitted to the master, and all temporary tweaks are discarded.

"discard" is the point: I do not want to add the loop around the tests over and over again.
I just want to temporarily set `$testRunCount` from `1` to a suiting value in `AppVeyor.yml` or in a build script of such a feature branch.
AppVeyor variables do not work well because they are global.

> BTW: The build script seems to not set the exit code.

I guess it has not worked before the one week old c110ab60f1ab3075b3e8e8213a1990da698fd006, which I did not take notice of.
And `$host.SetShouldExit` seems to not exit at once, which confused me.

Comment (RussKie) [reply appveyor.yml]: > "discard" is the point: I do not want to add the loop around the tests over and over again.
> I just want to temporarily set `$testRunCount` from `1` to a suiting value in `AppVeyor.yml` or in a build script of such a feature branch.
> AppVeyor variables do not work well because they are global.

This comes down to the point I'm trying to make - master or release/* branches are not the right place to test things.
We can have an `experimental` or `test` branch off the main trunk where appveyor.yml can be modified to whatever state maybe necessary, and have env variables that dynamically alter its behaviours.

Comment (RussKie) [reply appveyor.yml]: We can create a separate build project to build a specific branch (e.g. `experimental`) that would use the custom yml file.

File: appveyor.yml [hunk 8]
```diff
@@ -119,3 +75,28 @@ artifacts:
...
+            $total  = $result[1].Value
+            $passed = $result[4].Value
+            $failed = $result[5].Value
+            $notPassed = $total - $passed
+            $testErrors = $failed
+            $testName = $resultFilePath.Split('\\')[-3]
+
+            if ($notPassed -ne 0) {
+                $logFile = Get-ChildItem -recurse artifacts\$testName*.log
+                if ($testErrors -ne 0) { Push-AppveyorArtifact $logFile }
```
Comment (mstv) [inline appveyor.yml]: The XML file contains the most complete information in one place. The `.log` file is much more readable but scattered, especially console output. The AppVeyor tab "Tests" has issues (I don't trust it).

```suggestion
            if ($notPassed -ne 0) {
                $artifactName = "TestResult.${testErrors}err.${testName}"
                $logFile = Get-ChildItem -recurse artifacts\$testName*.log
                Push-AppveyorArtifact $logFile -FileName "${artifactName}.log"
                Push-AppveyorArtifact $resultFilePath -FileName "${artifactName}.xml"
```

File: appveyor.yml (hunk 6 above)
Comment (mstv) [reply appveyor.yml]: I am not convinced. It _will_ get out of sync. But go ahead. If we are lucky, we won't need it anymore.

File: appveyor.yml [hunk 9]
```diff
@@ -1,23 +1,15 @@
...
-  ARCHIVE_WITH_PDB: TRUE
-  GE_TEST_SLEEP_SECONDS_ON_HANG: 0
-  GE_TEST_LAUNCH_DEBUGGER_ON_HANG: 0
-  TEST_RUN_COUNT: 1
-  IGNORE_TEST_RESULT: FALSE
+matrix:
+  fast_finish: true
+
+# https://www.appveyor.com/docs/build-environment/#build-worker-images
+image: Visual Studio 2019
```
Comment (mstv) [inline appveyor.yml]: The following two variables are evaluated by the current code, i.e. should be kept.

```suggestion
image: Visual Studio 2019

environment:
  # The following entries define the default values of the environment variables.
  # They can be overridden in AppVeyor | Settings | Environment.
  GE_TEST_SLEEP_SECONDS_ON_HANG: 0
  GE_TEST_LAUNCH_DEBUGGER_ON_HANG: 0
```

**Comment by RussKie** [comment]
I create a new build project: https://ci.appveyor.com/project/gitextensions/gitextensions-experimental that builds `experimental/*` branches using `appveyor.experimental.yml`. You're welcome to modify the yml as you see necessary for testing purposes. The main build project will ignore both the yml and the experimental branches.
This should make moving changes between branches seamless.

**Comment by RussKie** [comment]
I have made an update, please check

File: appveyor.experimental.yml [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    for ($testRun = 1; $testRun -le $testRunCount; $testRun++)
+    {
+      if (!$env:RUN_UNIT_TESTS) {
+        & .\cibuild.cmd -test
+        if ($LastExitCode -ne 0  -and !$env:FAIL_IF_TEST_FAIL) { $host.SetShouldExit($LastExitCode);  }
+      }
+
+      if (!$env:RUN_INTEGRATION_TESTS) {
+        & .\cibuild.cmd -integrationTest
+        if ($LastExitCode -ne 0  -and !$env:FAIL_IF_TEST_FAIL) { $host.SetShouldExit($LastExitCode);  }
```
Comment (mstv) [inline appveyor.experimental.yml]: I think it should be:
```suggestion
      if ($env:RUN_UNIT_TESTS) {
        & .\cibuild.cmd -test
        if ($LastExitCode -ne 0  -and $env:FAIL_IF_TEST_FAIL) { $host.SetShouldExit($LastExitCode);  }
      }

      if ($env:RUN_INTEGRATION_TESTS) {
        & .\cibuild.cmd -integrationTest
        if ($LastExitCode -ne 0  -and $env:FAIL_IF_TEST_FAIL) { $host.SetShouldExit($LastExitCode);  }
```

**Comment by mstv** [comment]
Thank you.
